src/core/thread/network_data_service.hpp

```diff
@@ -324,42 +333,78 @@ class Manager : public InstanceLocator, private NonCopyable
     {
     public:
         OT_TOOL_PACKED_BEGIN
-        struct ServiceData
+        class AddrData
         {
         public:
-            explicit ServiceData(const Ip6::Address &aAddress, uint16_t aPort)
-                : mServiceNumber(kDnsSrpUnicastServiceNumber)
-                , mAddress(aAddress)
+            static constexpr uint8_t kMinLength = sizeof(Ip6::Address) + sizeof(uint16_t); // Address and port.
+
+            AddrData(const Ip6::Address &aAddress, uint16_t aPort, uint8_t aVersion)
+                : mAddress(aAddress)
                 , mPort(BigEndian::HostSwap16(aPort))
+                , mVersion(aVersion)
             {
-                OT_UNUSED_VARIABLE(mServiceNumber);
             }
 
+            uint8_t             GetLength(void) const { return (mVersion == 0) ? kMinLength : sizeof(AddrData); }
```

Comment:
This doesn't hold in general - it depends on usage.

* If, in the future, I still use version 0 and additional data gets added to the Unicast Dataset (unrelated to version), then it becomes necessary to explicitly encode version=0 in the dataset. While the length is higher than `kMinLength`.
* A Thread device could receive entries sent by peers/others that do contain an explicit version=0 value, so these entries use longer length. (However, maybe the GetLength function isn't relevant for these received entries? But that may not be obvious from this public interface.)

Maybe some static assert should be added that makes this compilation fail in case someone extends AddrData with more fields? Plus make clear that GetLength represents a "minimal encoding length" of the entry and not the actual length of an entry that was received from others?

Reply:
`GetLength()` is mainly used when appending this entry to Network Data. It represents the length we would use today when encoding this entry. It is not generally used for parsing, but yes, when parsing entries, if we use the `GetLength()` method, it should be treated as the minimum expected length. We use a similar model for TLV construction and parsing.

If we add additional (perhaps optional) fields to the entry in the future, the parsing code should be updated. Today it is the following (which performs multiple length checks before reading the corresponding fields that it knows are present). The parsing code typically should not use the `GetLength()` method.

```c++
Error Manager::DnsSrpUnicast::AddrData::ParseFrom(const uint8_t *aData, uint8_t aLength, DnsSrpUnicastInfo &aInfo)
{
    Error          error   = kErrorNone;
    const AddrData *addrData = reinterpret_cast<const AddrData *>(aData);

    VerifyOrExit(aLength >= kMinLength, error = kErrorParse);

    aInfo.mSockAddr.SetAddress(addrData->GetAddress());
    aInfo.mSockAddr.SetPort(addrData->GetPort());
    aInfo.mVersion = (aLength >= sizeof(AddrData)) ? addrData->GetVersion() : 0;

exit:
    return error;
}
```

I am not sure whether we can effectively safeguard against future changes (e.g., by adding assertions to the code today). If new fields are defined, a future PR adding support for them should handle them when parsing (and ensure backward compatibility).

Reply:
We could already safeguard this by generalizing the code now, as follows:

```
            uint8_t             GetLength(void) const {
                #if kMinLength + sizeof(uint8_t) == sizeof(AddrData)
                return (mVersion == 0) ? kMinLength : sizeof(AddrData);
                #else
                return sizeof(AddrData);
                #endif
            }
```
(so the future-developer doesn't need to remember to do this - this could be easily missed if someone extends fields)

Or else we can safeguard against future oversights, by just adding a single static check:
```
static_assert(kMinLength + sizeof(uint8_t) == sizeof(AddrData) , "Extending AddrData requires updating the term mVersion==0")
```

So it should be definitely possible I think. I'm just assuming some developer could be extending fields in AddrData without necessarily rewriting all the related methods, leading to subtle bugs later on. But maybe that would not happen as long as you're authoring it :)

Reply:
Thanks @EskoDijk  👍. I do like the idea of adding `static_assert(kMinLength + sizeof(uint8_t) == sizeof(AddrData))` to help make it future-proof. Unfortunately, such developer errors can happen often.

I added the following in the new push (for `AddrData` and also for `Service/ServerData`):

```cpp
static_assert(AddrData::kMinLength + sizeof(uint8_t) == sizeof(AddrData),
              "Update all methods/constants if adding new (optional) fields to `AddrData`.");
```